### ISSUE TITLE
[REVIEW]: PDLSM-FEM: Solver of Coupled Peridynamics Least Squares Minimization with Finite Element Method

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3668]
**Submitting author:** @QibangLiu  (<a href="http://orcid.org/0000-0001-7935-7907">Qibang Liu</a>)
**Repository:** <a href="https://github.com/QibangLiu/PDLSM-FEM_mpi" target ="_blank">https://github.com/QibangLiu/PDLSM-FEM_mpi</a>
**Version:** v1.0
**Editor:** @prashjha
**Reviewer:** @TLCFEM, @karthikncsu, @Balaje
**Archive:** <a href="https://doi.org/10.5281/zenodo.5784823" target="_blank">10.5281/zenodo.5784823</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/e429ae33fb3c06beb9a315f22354262f/status.svg)](https://joss.theoj.org/papers/e429ae33fb3c06beb9a315f22354262f)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/e429ae33fb3c06beb9a315f22354262f"><img src="https://joss.theoj.org/papers/e429ae33fb3c06beb9a315f22354262f/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/e429ae33fb3c06beb9a315f22354262f/status.svg)](https://joss.theoj.org/papers/e429ae33fb3c06beb9a315f22354262f)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/QibangLiu/PDLSM-FEM_mpi" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@TLCFEM & @karthikncsu & @Balaje, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @prashjha know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @TLCFEM
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/QibangLiu/PDLSM-FEM_mpi">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/QibangLiu/PDLSM-FEM_mpi">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@QibangLiu ) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @karthikncsu
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/QibangLiu/PDLSM-FEM_mpi">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/QibangLiu/PDLSM-FEM_mpi">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@QibangLiu ) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] #3765
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @Balaje
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/QibangLiu/PDLSM-FEM_mpi">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/QibangLiu/PDLSM-FEM_mpi">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@QibangLiu ) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] #3810
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @TLCFEM, @karthikncsu, @Balaje it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 556

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- doi.org/10.1016/S0022-5096(99)00029-0 is OK
- 10.1007/s10659-007-9125-1 is OK
- 10.1007/s00466-017-1469-1 is OK
- 10.1016/j.cma.2014.06.016 is OK
- 10.1016/j.cma.2019.01.032 is OK
- 10.1016/j.engfracmech.2020.107483 is OK
- 10.1007/s10704-012-9745-8 is OK
- 10.1007/978-1-4614-8465-3 is OK
- 10.1016/j.engfracmech.2021.107862 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.12 s (533.5 files/s, 117414.9 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
C++                             30            930           2024          10106
C/C++ Header                    29            129            166            880
TeX                              1             10              0            110
CMake                            4             13             36             81
Markdown                         2              9              0             32
-------------------------------------------------------------------------------
SUM:                            66           1091           2226          11209
-------------------------------------------------------------------------------


Statistical information for the repository '6da3ee72f6882291fd61d3e9' was
gathered on 2021/08/29.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
QibangLiu                       21         16580           2345          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
QibangLiu                 14235           85.9          2.2               20.79
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03668/joss.03668/10.21105.joss.03668.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03668/joss.03668/10.21105.joss.03668.pdf) :page_facing_up: :point_left:

**Comment by prashjha** [comment]
Thank you, @TLCFEM, @karthikncsu, @Balaje for agreeing to be reviewers. Please read the first couple of comments in this thread and also reviewer guidelines [here](https://joss.readthedocs.io/en/latest/reviewer_guidelines.html). Also, you can browse the closed "REVIEW" issues on the "joss-reviews" repository to get some ideas on how to complete the reviews. Thanks again and good luck!

**Comment by whedon** [comment]
:wave: @karthikncsu, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @Balaje, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @TLCFEM, please update us on how your review is going (this is an automated reminder).

**Comment by prashjha** [comment]
Hello @Balaje @TLCFEM, I am wondering how is the review going? 

@karthikncsu please let me know when you can start the review. If you are unable to perform the review, let me know.

**Comment by karthikncsu** [comment]
@prashjha, Sorry for the delay. I have started the review!

**Comment by prashjha** [comment]
@karthikncsu great. thanks!! :)

**Comment by TLCFEM** [comment]
working on it.

**Comment by karthikncsu** [comment]
@QibangLiu

Installation of the PDLSM-FEM solver software requires setting up the paths for MKLROOT and MPIROOT as mentioned in the manual: page 7 section 2.2 Build on Linux

export MKLROOT =/ homes / MKL_2020 / compilers_and_libraries_2020 .1.217/ linux / mkl
export MPIROOT =/ opt / software / OpenMPI /4.0.5 - iccifort -2020.4.304

I suppose these pathnames are available in the older version of intel parallel studio.

Intel had discontinued the old intel parallel studio product and have released the open-source oneAPI tool kit, replacing Intel Parallel Studio XE on December 8, 2020.

https://software.intel.com/content/www/us/en/develop/tools/oneapi/all-toolkits.html
https://en.wikipedia.org/wiki/Intel_Parallel_Studio

I faced similar issues recently when my intel parallel studio license has expired and had to change my code to be compatible with oneAPI. I am not able to install the package as I don't have the intel parallel studio license. Please modify the code and the documentation with the new paths for the mkl and iccifort filenames in the open API installation directory.

Please let me know if there is an alternate way of installation of the software.

**Comment by QibangLiu** [comment]
@karthikncsu  Thanks for your comments. I didn't realize they closed the old version and released the oneAPI toolkit. 
The principles to set environment variables are that:
 (a) Under the path "MKLROOT/include", there is the "mkl.h file", and under the path "MKLROOT/lib/intel64", there are libraries as the MKL link line advisor show, such as "libmkl_intel_ilp64.so", "libmkl_core.so".
(b) Under the path "MPIROOT/include", there is the header file "mpi.h",  and under the path "MPIROOT/lib", there is the library "libmpi.so".
This explanation is added in the manual file. 

For the MPIROOT, the OpenMPI is installed under the path "/opt / software / OpenMPI /4.0.5 - iccifort -2020.4.304" on my machine, but not installed with the intel parallel studio product.
Anyway, I modified the values based on the oneAPI toolkit setting as your request:
export MKLROOT=/homes/qibangliu/intel/oneapi/mkl/2021.3.0
export MPIROOT=/homes/qibangliu/intel/oneapi/mpi/2021.3.0

I noted that the file "libmpi.so" is under the path /homes/qibangliu/intel/oneapi/mpi/2021.3.0/lib/release, so the 
HINTS value in the FindMPI.cmake file should be modified as:
find_library(MPI_LIBRARY NAMES libmpi.so  HINTS $ENV{MPIROOT}/lib $ENV{MPIROOT}/lib/release)

The updates were posted on GitHub.

**Comment by QibangLiu** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03668/joss.03668/10.21105.joss.03668.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03668/joss.03668/10.21105.joss.03668.pdf) :page_facing_up: :point_left:

**Comment by QibangLiu** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- doi.org/10.1016/S0022-5096(99)00029-0 is OK
- 10.1007/s10659-007-9125-1 is OK
- 10.1007/s00466-017-1469-1 is OK
- 10.1016/j.cma.2014.06.016 is OK
- 10.1016/j.cma.2019.01.032 is OK
- 10.1016/j.engfracmech.2020.107483 is OK
- 10.1007/s10704-012-9745-8 is OK
- 10.1007/978-1-4614-8465-3 is OK
- 10.1016/j.engfracmech.2021.107862 is OK
- 10.1007/s42102-021-00060-3 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by karthikncsu** [comment]
Hi @QibangLiu  Thank you for updating the manual with the settings to run the PDLSM-FEM with one API.

I tried to install on my ubuntu system and it returned below error:

/opt/intel/oneapi/mkl/2021.3.0//lib/intel64/libmkl_gnu_thread.so: undefined reference to `GOMP_loop_end_nowait'
/opt/intel/oneapi/mkl/2021.3.0//lib/intel64/libmkl_gnu_thread.so: undefined reference to `GOMP_loop_ordered_static_start'
/opt/intel/oneapi/mkl/2021.3.0//lib/intel64/libmkl_gnu_thread.so: undefined reference to `omp_get_thread_num'
/opt/intel/oneapi/mkl/2021.3.0//lib/intel64/libmkl_gnu_thread.so: undefined reference to `GOMP_loop_dynamic_next'
/opt/intel/oneapi/mkl/2021.3.0//lib/intel64/libmkl_gnu_thread.so: undefined reference to `GOMP_sections_next'
/opt/intel/oneapi/mkl/2021.3.0//lib/intel64/libmkl_gnu_thread.so: undefined reference to `GOMP_parallel_end'
/opt/intel/oneapi/mkl/2021.3.0//lib/intel64/libmkl_gnu_thread.so: undefined reference to `omp_get_num_threads'
/opt/intel/oneapi/mkl/2021.3.0//lib/intel64/libmkl_gnu_thread.so: undefined reference to `omp_set_max_active_levels'
/opt/intel/oneapi/mkl/2021.3.0//lib/intel64/libmkl_gnu_thread.so: undefined reference to `GOMP_parallel_sections_start'
collect2: error: ld returned 1 exit status
CMakeFiles/PDLSM-FEM.dir/build.make:854: recipe for target 'PDLSM-FEM' failed
make[2]: *** [PDLSM-FEM] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/PDLSM-FEM.dir/all' failed
make[1]: *** [CMakeFiles/PDLSM-FEM.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2

Let me know a way around to resolve this. I would also recommend you to provide requirements files in the Github repo.

**Comment by TLCFEM** [comment]
need flag -fopenmp

**Comment by QibangLiu** [comment]
@karthikncsu @TLCFEM The error is because the library "libmkl_gnu_thread.so" requires libgomp. I have tried to use combination of OpenMP and MPI to parallel the code. It is my bad that I forgot to remove the requirement of libmkl_gnu_thread. 

I have replaced the libmkl_gnu_thread.so by libmkl_sequential.so. 

I tried to extract the required files, but some of them have very large sizes, that are difficult to post on Github. 

By the way, if your use the MPI version in the OneAPI package, you may have to replace the requirement library "libmkl_blacs_openmpi_ilp64.so" by "libmkl_blacs_intelmpi_ilp64.so", following the instruction in the manual file (the head of section 2), as the MPI in the OneAPI package is "Intel MPI", but the file "FindMKL.cmake" set it for OpenMPI.

Updates are posted on Github.

**Comment by Balaje** [comment]
Hi @prashjha, Sorry for the delay. I have started the review and working on it. Hope to finish by the end of next week! Thanks.

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03668/joss.03668/10.21105.joss.03668.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03668/joss.03668/10.21105.joss.03668.pdf) :page_facing_up: :point_left:

**Comment by QibangLiu** [comment]
@Balaje BTW, we first proposed the coupled PDLSM-FEM model, and I don't think there are some similar software packages.

**Comment by TLCFEM** [comment]
Hi all, any progress regarding this paper? Wondering if there is anything I can do to push this forward a bit.

**Comment by prashjha** [comment]
@TLCFEM I am waiting on @Balaje and @karthikncsu for their decision. Thanks for asking.

**Comment by prashjha** [comment]
@Balaje and @karthikncsu how far are you with the reviews? No hurry though...your time is sincerely appreciated. :)

**Comment by TLCFEM** [comment]
Just to confirm, is there anything I shall do apart from the checklist? Any review document shall be provided?

**Comment by prashjha** [comment]
@TLCFEM that will be all but if you have comments or any point on other aspects not covered in the checklist, you could let the authors know. As far as I know, no review document is required from your and our side as the checklist and issues are records of review.

**Comment by prashjha** [comment]
Hi @Balaje and @karthikncsu, could you please provide an update on how the review is going and a tentative time to finish it? I truly appreciate your efforts. :)

**Comment by karthikncsu** [comment]
Hi @prashjha, I have completed the review. Please let me know if I need to do anything else.

**Comment by QibangLiu** [comment]
whedon generate pdf

**Comment by QibangLiu** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03668/joss.03668/10.21105.joss.03668.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03668/joss.03668/10.21105.joss.03668.pdf) :page_facing_up: :point_left:

**Comment by Balaje** [comment]
Hi @prashjha, I have completed the review. I am ticking off all the boxes. Thanks.

**Comment by prashjha** [comment]
@Balaje @karthikncsu @TLCFEM thank you all for your efforts. I am going to look at the draft one more time and soon hand it over to EiC for final acceptance.

**Comment by prashjha** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03668/joss.03668/10.21105.joss.03668.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03668/joss.03668/10.21105.joss.03668.pdf) :page_facing_up: :point_left:

**Comment by prashjha** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- doi.org/10.1016/S0022-5096(99)00029-0 is OK
- 10.1007/s10659-007-9125-1 is OK
- 10.1007/s00466-017-1469-1 is OK
- 10.1016/j.cma.2014.06.016 is OK
- 10.1016/j.cma.2019.01.032 is OK
- 10.1016/j.engfracmech.2020.107483 is OK
- 10.1007/s10704-012-9745-8 is OK
- 10.1007/978-1-4614-8465-3 is OK
- 10.1016/j.engfracmech.2021.107862 is OK
- 10.1007/s42102-021-00060-3 is OK
- 10.1016/j.engfracmech.2009.02.016 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by prashjha** [comment]
Hi @QibangLiu, could you address these points:

1. Draft corrections (suggestions):

- Line 7: I will suggest removing single quote around PDLSM-FEM solver
- Line 8: "and 3D using MPI parallelism" instead of "3D by MPI technique"?
- Line 8, 9: "This cross-platform solver is written in a C++ language and includes ..."?
- Line 27: "that of a PD with a spherical neighborhood"?
- Line 27: "However, their PD ..."?
- Line 37: you have used "much" multiple times and further "much" sounds not so scientific. How about "... mesh relatively fast and utilizes less memory compared to pure PD model implementations."?
- Line 39: In "In many published PD...", please specify references as "many" does not sound right.
- Line 44: "... in an easy way similar to FEM without introducing fictitious layer"?
- Line 44: "... solver enables two criteria ..."?
- Line 45: In "for 2D crack ...", is it just for 2D or is it is true also for 3D?
- Line 49: "It can also be ..."?

2. Once you have updated the draft, can you also do a 'tagged' release of your code, and

3. archive the release using zenedo or other methods.

Once you are done, I will run a few commands and hand your paper to EiC for the final decision.

**Comment by QibangLiu** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03668/joss.03668/10.21105.joss.03668.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03668/joss.03668/10.21105.joss.03668.pdf) :page_facing_up: :point_left:

**Comment by prashjha** [comment]
@whedon set 10.5281/zenodo.5784823 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.5784823" target="_blank">10.5281/zenodo.5784823</a> is the archive.

**Comment by prashjha** [comment]
Thanks, @QibangLiu. One last thing: could you change the title in zenodo archive to match the title of joss paper?

**Comment by QibangLiu** [comment]
@prashjha DONE.

**Comment by prashjha** [comment]
Great...handing it to EiC!!

**Comment by prashjha** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- doi.org/10.1016/S0022-5096(99)00029-0 is OK
- 10.1007/s10659-007-9125-1 is OK
- 10.1007/s00466-017-1469-1 is OK
- 10.1016/j.cma.2014.06.016 is OK
- 10.1016/j.cma.2019.01.032 is OK
- 10.1016/j.engfracmech.2020.107483 is OK
- 10.1007/s10704-012-9745-8 is OK
- 10.1007/978-1-4614-8465-3 is OK
- 10.1016/j.engfracmech.2021.107862 is OK
- 10.1007/s42102-021-00060-3 is OK
- 10.1016/j.engfracmech.2009.02.016 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2830

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2830, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1471671700626542594) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2833
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03668](https://doi.org/10.21105/joss.03668)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by kyleniemeyer** [comment]
Congratulations @QibangLiu on your article's publication in JOSS!

Many thanks to @TLCFEM, @karthikncsu, and @Balaje for reviewing this submission, and @prashjha for editing.

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03668/status.svg)](https://doi.org/10.21105/joss.03668)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03668">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03668/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03668/status.svg
   :target: https://doi.org/10.21105/joss.03668
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03668/status.svg)](https://doi.org/10.21105/joss.03668)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
